contracts/IdentityManager.sol

```diff
@@ -0,0 +1,162 @@
+  // Note they also have to change the owner of the Proxy over to this, but after calling this
+  function registerIdentity(address owner, address recoveryKey) {
+    if (recoveryKey == address(0)) throw;
+    if (owners[msg.sender][owner] > 0 || recoveryKeys[msg.sender] > 0 ) throw; // Deny any funny business
+    owners[msg.sender][owner] = now - adminTimeLock; // This is to ensure original owner has full power from day one
+    recoveryKeys[msg.sender] = recoveryKey;
+    IdentityCreated(msg.sender, msg.sender, owner, recoveryKey);
+  }
+
+  // Primary forward function
+  function forwardTo(Proxy identity, address destination, uint value, bytes data) onlyOwner(identity) {
+    identity.forward(destination, value, data);
+  }
+
+  // an owner can add a new device instantly
+  function addOwner(Proxy identity, address newOwner) onlyOlderOwner(identity) rateLimited(identity) {
+    owners[identity][newOwner] = now;
+    OwnerAdded(identity, newOwner, msg.sender);
+  }
+
+  // a recovery key owner can add a new device with 'adminRate' wait time
+  function addOwnerFromRecovery(Proxy identity, address newOwner) onlyRecovery(identity) rateLimited(identity) {
+    if (owners[identity][newOwner] > 0) throw;
+    owners[identity][newOwner] = now;
+    OwnerAdded(identity, newOwner, msg.sender);
+  }
+
+  // an owner can remove another owner instantly
+  function removeOwner(Proxy identity, address owner) onlyOlderOwner(identity) rateLimited(identity) {
+    owners[identity][owner] = 0;
+    OwnerRemoved(identity, owner, msg.sender);
+  }
+
+  // an owner can change the recoverykey whenever they want to
+  function changeRecovery(Proxy identity, address recoveryKey) onlyOlderOwner(identity) rateLimited(identity) {
+    if (recoveryKey == address(0)) throw;
+    recoveryKeys[identity] = recoveryKey;
+    RecoveryChanged(identity, recoveryKey, msg.sender);
+  }
+
+  // an owner can migrate away to a new IdentityManager
+  function initiateMigration(Proxy identity, address newIdManager) onlyOlderOwner(identity) {
+    migrationInitiated[identity] = now;
+    migrationNewAddress[identity] = newIdManager;
+    MigrationInitiated(identity, newIdManager, msg.sender);
+  }
+
+  // any owner can cancel a migration
+  function cancelMigration(Proxy identity) onlyOwner(identity) {
+    address canceledManager = migrationNewAddress[identity];
+    migrationInitiated[identity] = 0;
+    migrationNewAddress[identity] = 0;
+    MigrationCanceled(identity, canceledManager, msg.sender);
+  }
+
+  // owner needs to finalize migration once adminTimeLock time has passed
+  // WARNING: before transfering to a new address, make sure this address is "ready to recieve" the proxy.
+  // Not doing so risks the proxy becoming stuck.
+  function finalizeMigration(Proxy identity) onlyOlderOwner(identity) {
+    if (migrationInitiated[identity] > 0 && migrationInitiated[identity] + adminTimeLock < now) {
```

Comment:
Explanatory comment on my above comment. If `migrationNewAddress[identity] == address(0)` here we could loose an identity completely. But the right place to check for it is in `initiateMigration`